Ожидает проверки

WorldOfBatle opened the pull request.

WorldOfBattle committed:
Создана заготовка проекта с функциями

WorldOfBattle committed:
В task1() считываем n и x, проверяем n

WorldOfBattle committed:
Добавлен двойной цикл для подсчёта суммы sin-sin-...-sin

WorldOfBattle committed:
Вывод итогового значения S

WorldOfBattle committed:
Считываем k и x, проверяем k

WorldOfBattle committed:
Добавлен цикл для вычисления произведения P и вывод результата

WorldOfBattle committed:
Считываем eps, проверяем > 0

WorldOfBattle committed:
Организован цикл суммирования, остановка когда term < eps, а также подсчёт количества итераций. Вывод результата и количества итераций.

WorldOfBattle committed:
Считываем a,b,eps, печатаем заголовок таблицы

WorldOfBattle committed:
Реализована вспомогательная функция getTerm(n, x) для Tn, по условиям задачи №4

WorldOfBattle committed:
Реализован цикл по x, суммирование ряда до eps, вывод в таблицу. Финал.

tak333 commented on HW-5/main.cpp:
```diff
@@ -0,0 +1,225 @@
+        // Внутренний цикл: j от 1 до i
+        for (int j = 1; j <= i; j++)
+        {
+            val = std::sin(val);
```
Неэффективно, надо использовать рекуррентные отношения.

tak333 commented on HW-5/main.cpp:
```diff
@@ -0,0 +1,225 @@
+    for (int n = 0; n <= k; n++)
+    {
+        // Вычисляем знак: (-1)^(n-1)
+        double sign = std::pow(-1.0, double(n - 1));
```
То же замечание: неэффективно.

tak333 commented on HW-5/main.cpp:
```diff
@@ -0,0 +1,225 @@
+        double sign = std::pow(-1.0, double(n - 1));
+
+        // x^(2n)
+        double xPow = std::pow(x, 2.0 * n);
```
Здесь и во всех подобных местах.

tak333 commented on HW-5/main.cpp:
```diff
@@ -0,0 +1,225 @@
+    while (true)
+    {
+        // Вычисляем i-й член: 1 / sqrt(3^i) = 1 / 3^(i/2)
+        double term = 1.0 / std::pow(3.0, i / 2.0);
```
То же.

tak333 commented on HW-5/main.cpp:
```diff
@@ -0,0 +1,225 @@
+    {
+        // n >= 3
+        // 1) Вычисляем x^(2n-1)
+        double power = std::pow(x, 2.0 * n - 1.0);
```
То же.

tak333 commented on HW-5/main.cpp:
```diff
@@ -0,0 +1,225 @@
+
+        for (int evenVal = 2; evenVal <= 2 * (n - 1); evenVal += 2)
+        {
+            denProd *= evenVal;
```
И тут.

tak333 commented on HW-5/main.cpp:
```diff
@@ -0,0 +1,225 @@
+            sum += Tn;
+
+            // ограничимся, скажем, 200 итерациями, чтобы не застрять
+            if (count > 200)
```
Вопрос: по какой причине мы можем застрять?

WorldOfBatle replied:
Ограничение по итерациям введено для того, чтобы избежать бесконечного цикла, если ряд сходится медленно или точность ε выбрана слишком маленькой.

tak333 reviewed (changes requested):
Совсем не используете рекуррентные отношения, а надо: на С++ мы думаем о производительности, иначе зачем это всё :,)

Рассчитывайте каждый новый член на базе предыдущего, а не начинайте просчёт на каждом шаге с нуля.

WorldOfBattle committed:
Исправлен код согласно комментариям:
– применены рекуррентные соотношения для вычисления значений (sin, степени и т.д.) вместо повторного вызова std::pow;
– обновлён расчёт знака в task2 через переключение переменной;
– оптимизировано вычисление ряда в task3 и вычисление Tn в task4;
– добавлено объяснение ограничения по итерациям (200) для защиты от бесконечного цикла.
